README.md
# **Interactive Corporea**
W.I.P.

changelog.txt
Build: 1.16.5-36.2.22 - Tue Dec 14 20:03:13 GMT 2021
	jaredlll08:
		(1.16) Fixed KeyMappings only checking if they conflict with themselves. (#8257)

=========
Build: 1.16.5-36.2.21 - Tue Dec 14 19:50:38 GMT 2021
	pupnewfster:
		Backport #8254 to 1.16.5: Add RenderArmEvent to make overriding just the arm rendering not require copying nearly as much vanilla code (#8299)

=========
Build: 1.16.5-36.2.20 - Fri Dec 10 10:00:01 GMT 2021
	oriondevelopment:
		[CVE-2021-44228]: Update Log4J to fix the security issue inside it (#8270)

=========
Build: 1.16.5-36.2.19 - Wed Nov 24 02:14:10 GMT 2021
	daemonumbra:
		Fix potential issue with BackgroundScanHandler for users with slow disc IO (#8211)

=========
Build: 1.16.5-36.2.18 - Mon Nov 22 21:19:17 GMT 2021
	pupnewfster:
		Backport #8207 to 1.16 (#8209)

=========
Build: 1.16.5-36.2.17 - Mon Nov 22 20:46:52 GMT 2021
	xfacthd:
		[1.16.x] Add ForgeSpawnEggItem to lazily handle EntityTypes (#8150)
		
		* Add ForgeSpawnEggItem to lazily handle EntityTypes
		
		* Remove OnlyIn
		
		* Surround EntityType#spawn() call with try/catch

=========
Build: 1.16.5-36.2.16 - Sat Nov 20 15:45:11 GMT 2021
	Unbekannt1998:
		change the AdvancementProvider to make it possabible for mods to hook into it (#7488)
		
		* back Port changes from #8120

=========
Build: 1.16.5-36.2.15 - Sat Nov 20 15:33:16 GMT 2021
	oriondevelopment:
		Implements Lazy ItemStack Capabilities (#7879)

=========
Build: 1.16.5-36.2.14 - Sat Nov 20 14:52:41 GMT 2021
	xfacthd:
		Pass IModelData through IForgeBakedModel#getModelData() before asking the IBakedModel for the particle texture (#8107)

=========
Build: 1.16.5-36.2.13 - Sat Nov 20 14:30:21 GMT 2021
	xfacthd:
		[1.16.x] Fix blocks with flammable material catching fire from lava when they are marked as non-flammable (#8161)

=========
Build: 1.16.5-36.2.12 - Sat Nov 20 14:17:06 GMT 2021
	xfacthd:
		[1.16] Add accessor to an unmodifiable view of the fullPots map in FlowerPotBlock (#8109)
		
		* Add accessor to an unmodifiable view of the fullPots map in FlowerPotBlock
		
		* Fix non-empty pot check
		
		* Remove exception from getter and add test mod

=========
Build: 1.16.5-36.2.11 - Sat Nov 20 13:56:21 GMT 2021
	championash5357:
		Ensure that all render targets in a PostChain use the same format (#8138)
		
		Fixes #6995 

=========
Build: 1.16.5-36.2.10 - Sat Nov 20 13:42:51 GMT 2021
	championash5357:
		Make Broken Blocks Display Associated FluidState (#8129)
		
		Fixes #7253 for 1.16.x

=========
Build: 1.16.5-36.2.9 - Thu Nov 04 07:09:52 GMT 2021
	gigaherz:
		Implement GUI Stacking feature. (#6927)
		
		* Implement GUI Stacking feature.
		You can now use `mc.pushGuiLayer` and `mc.popGuiLayer` to manage the layers.
		mc.displayGuiScreen behaves such that if called with a non-null screen it replaces the entire stack, and if called with null it closes the entire stack.
		
		* Apply review comment fixes from 1.17 counterpart.
		
		* Revert unintended changes.
		
		* Missed these reverts.

=========
Build: 1.16.5-36.2.8 - Sun Oct 03 16:51:54 GMT 2021
	sciwhiz12:
		Add workaround for pairs of `'` being stripped (#8074)
		
		Backport of #8050

	ititus:
		Fix typo in FluidBlockWrapper (#7880)

=========
Build: 1.16.5-36.2.6 - Mon Sep 27 20:24:54 GMT 2021
	brandon4261:
		Backport MC-209819 (#7923)
		
		Backport fix for mc bug https://bugs.mojang.com/browse/MC-209819

=========
Build: 1.16.5-36.2.5 - Sun Sep 19 23:50:14 GMT 2021
	zml:
		[1.16.5] Bump Mixin to 0.8.4 (#8098)

=========
Build: 1.16.5-36.2.4 - Tue Aug 31 16:32:40 GMT 2021
	djbake101:
		[1.16.5] Backport OnDatapackSyncEvent from 1.17.1 (#7985)

=========
Build: 1.16.5-36.2.3 - Tue Aug 31 15:35:00 GMT 2021
	sciwhiz12:
		[1.16.x] Fix all layers of ItemLayerModel being fullbright (#8040)
		
		* Fix all layers of ItemLayerModel being fullbright
		
		Backport of #8039
		
		* Remove duplicate test resources
		
		These resources have corresponding equivalents in the generated tests
		resources folder.

=========
Build: 1.16.5-36.2.2 - Mon Jul 26 21:37:03 GMT 2021
	onelemonyboi:
		Fix rounding related errors with small entity pathfinding. Closes #7730

=========
Build: 1.16.5-36.2.1 - Sat Jul 24 17:39:23 GMT 2021
	supermartijn642:
		Pass the 'includeName' flag to the ItemTossEvent (#7758)

=========
Build: 1.16.5-36.2.0 - Thu Jul 22 01:45:10 GMT 2021
	lexmanos:
		New 1.16.5 RB.

=========
Build: 1.16.5-36.1.66 - Wed Jul 21 21:21:25 GMT 2021
	sizableshrimp:
		ForgeGradle 5.1 and Gradle 7, which allows development on J16. Runtime still targets J8. (#7877)

=========
Build: 1.16.5-36.1.65 - Tue Jul 20 05:01:10 GMT 2021
	lexmanos:
		Work around javac scoping issue.

	lexmanos:
		Fix LAN worlds where registry entries have been removed.

=========
Build: 1.16.5-36.1.63 - Mon Jul 19 19:54:59 GMT 2021
	william:
		Fix FakePlayers having null `connection` and add FakePlayer test mod (#7733)

=========
Build:  - Sat Jul 17 16:23:44 GMT 2021
	ateranimavis:
		Use '{assets_root}' rather than task output path for userdev config (#7873)
		
		Signed-off-by: AterAnimAvis <AterAnimAvis@gmail.com>

=========
Build: 1.16.5-36.1.61 - Sat Jul 17 11:14:53 GMT 2021
	thelnfamous1:
		Patch several mob classes to let them use modded versions of their ranged weapons (#7845)

	baileyholl:
		Adds a damage double in Lightning Entity and patches Entity to use it. (#7849)

	thesilkminer:
		Fix missing userdev config for datagen runs (#7871)
		
		Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

=========
Build: 1.16.5-36.1.58 - Sat Jul 17 02:45:36 GMT 2021
	awsomesauce239:
		Add Fishing bobber projectile impact event (#7842)

	onelemonyboi:
		Removes misleading Nullable annotation from canCreatureSpawn (#7840)

	onelemonyboi:
		Removes misapplied patch in Entity (#7841)

	onelemonyboi:
		Updates MouseHelper to allow for Horizontal Scroll to be recognized as Vertical Scroll (#7834)

	onelemonyboi:
		Corrects the assigning of mob kills, Fixes MC-121048 (#7836)

=========

src/main/java/shblock/interactivecorporea/common/util/Perlin.java
package shblock.interactivecorporea.common.util;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

/**
 * From: https://github.com/raubirius/GRobot/blob/master/kni%C5%BEnica/podpora/PerlinNoise.java
 * (Modified)
 */
public class Perlin {
//  private final static int[] permutation =
//      {
//          // Hash lookup table as defined by Ken Perlin. This is a randomly
//          // arranged array of all numbers from 0 – 255 inclusive.
//          151, 160, 137, 91, 90, 15, 131, 13, 201, 95, 96, 53, 194, 233, 7,
//          225, 140, 36, 103, 30, 69, 142, 8, 99, 37, 240, 21, 10, 23, 190, 6,
//          148, 247, 120, 234, 75, 0, 26, 197, 62, 94, 252, 219, 203, 117, 35,
//          11, 32, 57, 177, 33, 88, 237, 149, 56, 87, 174, 20, 125, 136, 171,
//          168, 68, 175, 74, 165, 71, 134, 139, 48, 27, 166, 77, 146, 158, 231,
//          83, 111, 229, 122, 60, 211, 133, 230, 220, 105, 92, 41, 55, 46, 245,
//          40, 244, 102, 143, 54, 65, 25, 63, 161, 1, 216, 80, 73, 209, 76, 132,
//          187, 208, 89, 18, 169, 200, 196, 135, 130, 116, 188, 159, 86, 164,
//          100, 109, 198, 173, 186, 3, 64, 52, 217, 226, 250, 124, 123, 5, 202,
//          38, 147, 118, 126, 255, 82, 85, 212, 207, 206, 59, 227, 47, 16, 58,
//          17, 182, 189, 28, 42, 223, 183, 170, 213, 119, 248, 152, 2, 44, 154,
//          163, 70, 221, 153, 101, 155, 167, 43, 172, 9, 129, 22, 39, 253, 19,
//          98, 108, 110, 79, 113, 224, 232, 178, 185, 112, 104, 218, 246, 97,
//          228, 251, 34, 242, 193, 238, 210, 144, 12, 191, 179, 162, 241, 81,
//          51, 145, 235, 249, 14, 239, 107, 49, 192, 214, 31, 181, 199, 106,
//          157, 184, 84, 204, 176, 115, 121, 50, 45, 127, 4, 150, 254, 138, 236,
//          205, 93, 222, 114, 67, 29, 24, 72, 243, 141, 128, 195, 78, 66, 215,
//          61, 156, 180
//      };
//
//  // Doubled permutation to avoid overflow.
  private static final Random SEED_GENERATOR = new Random();
  private static final Map<Long, int[]> P_CACHE = new HashMap<>();

  public int repeat;
  private int[] p = new int[512];

  public Perlin(long seed, int repeat) {
    this.repeat = repeat;
    setSeed(seed);
  }

  public Perlin() {
    repeat = -1;
    setSeed(SEED_GENERATOR.nextLong());
  }

  public Perlin(long seed) {
    repeat = -1;
    setSeed(seed);
  }

  public void setSeed(long seed) {
    Random random = new Random(seed);
    for (int x = 0; x < 256; ++x) {
      int n = random.nextInt(256);
      p[x] = n;
      p[x + 256] = n;
    }
    P_CACHE.put(seed, p);
  }


  public double octavePerlin(double x, double y, double z,
                             int octaves, double persistence) {
    double total = 0;
    double frequency = 1;
    double amplitude = 1;

    // Used to normalize the result to the range ⟨−1.0, 1.0⟩.
    double maxValue = 0;

    for (int i = 0; i < octaves; ++i) {
      total += perlin(x * frequency, y * frequency,
          z * frequency) * amplitude;

      maxValue += amplitude;

      amplitude *= persistence;
      frequency *= 2;
    }

    return total / maxValue;
  }

  // The noise within the range ⟨−1, 1⟩.
  public double perlin(double x, double y, double z) {
    if (repeat > 0) {
      // If we have any repeat on, change the coordinates to their
      // “local” repetitions.
      x = x % repeat;
      y = y % repeat;
      z = z % repeat;
    }

    // Calculate the “unit cube” that the point asked will be located in.
    // The left bound is (|_x_|, |_y_|, |_z_|) and the right bound is
    // that plus 1. Next we calculate the location (from 0.0 to 1.0) in
    // that cube. We also fade the location to smooth the result.

    int xi = (int) x & 255;
    int yi = (int) y & 255;
    int zi = (int) z & 255;

    int aaa = p[p[p[xi] + yi] + zi];
    int aba = p[p[p[xi] + inc(yi)] + zi];
    int aab = p[p[p[xi] + yi] + inc(zi)];
    int abb = p[p[p[xi] + inc(yi)] + inc(zi)];
    int baa = p[p[p[inc(xi)] + yi] + zi];
    int bba = p[p[p[inc(xi)] + inc(yi)] + zi];
    int bab = p[p[p[inc(xi)] + yi] + inc(zi)];
    int bbb = p[p[p[inc(xi)] + inc(yi)] + inc(zi)];

    double xf = x - (int) x;
    double yf = y - (int) y;
    double zf = z - (int) z;
    double u = fade(xf);
    double v = fade(yf);
    double w = fade(zf);

    // The gradient function calculates the dot product between
    // a pseudorandom gradient vector and the vector from the input
    // coordinate to the 8 surrounding points in its unit cube. This
    // is all then lerped together as a sort of weighted average
    // based on the faded (u, v, w) values we made earlier.

    double x1, x2;

    x1 = lerp(
        grad(aaa, xf, yf, zf),
        grad(baa, xf - 1, yf, zf), u);
    x2 = lerp(
        grad(aba, xf, yf - 1, zf),
        grad(bba, xf - 1, yf - 1, zf), u);
    double y1 = lerp(x1, x2, v);

    x1 = lerp(
        grad(aab, xf, yf, zf - 1),
        grad(bab, xf - 1, yf, zf - 1), u);
    x2 = lerp(
        grad(abb, xf, yf - 1, zf - 1),
        grad(bbb, xf - 1, yf - 1, zf - 1), u);
    double y2 = lerp(x1, x2, v);

    return lerp(y1, y2, w);
  }

  public double octavePerlinHalf(double x, double y, double z,
                                 int octaves, double persistence) {
    double total = 0;
    double frequency = 1;
    double amplitude = 1;

    // Used to normalize the result to the range ⟨0.0, 1.0⟩.
    double maxValue = 0;

    for (int i = 0; i < octaves; ++i) {
      total += perlinHalf(x * frequency, y * frequency,
          z * frequency) * amplitude;

      maxValue += amplitude;

      amplitude *= persistence;
      frequency *= 2;
    }

    return total / maxValue;
  }

  // This version bounds the output range to ⟨0, 1⟩ for convenience.
  public double perlinHalf(double x, double y, double z) {
    if (repeat > 0) {
      // If we have any repeat on, change the coordinates to their
      // “local” repetitions.
      x = x % repeat;
      y = y % repeat;
      z = z % repeat;
    }

    // Calculate the “unit cube” that the point asked will be located in.
    // The left bound is (|_x_|, |_y_|, |_z_|) and the right bound is
    // that plus 1. Next we calculate the location (from 0.0 to 1.0) in
    // that cube. We also fade the location to smooth the result.

    int xi = (int) x & 255;
    int yi = (int) y & 255;
    int zi = (int) z & 255;

    int aaa = p[p[p[xi] + yi] + zi];
    int aba = p[p[p[xi] + inc(yi)] + zi];
    int aab = p[p[p[xi] + yi] + inc(zi)];
    int abb = p[p[p[xi] + inc(yi)] + inc(zi)];
    int baa = p[p[p[inc(xi)] + yi] + zi];
    int bba = p[p[p[inc(xi)] + inc(yi)] + zi];
    int bab = p[p[p[inc(xi)] + yi] + inc(zi)];
    int bbb = p[p[p[inc(xi)] + inc(yi)] + inc(zi)];

    double xf = x - (int) x;
    double yf = y - (int) y;
    double zf = z - (int) z;
    double u = fade(xf);
    double v = fade(yf);
    double w = fade(zf);

    // The gradient function calculates the dot product between
    // a pseudorandom gradient vector and the vector from the input
    // coordinate to the 8 surrounding points in its unit cube. This
    // is all then lerped together as a sort of weighted average
    // based on the faded (u, v, w) values we made earlier.

    double x1, x2;

    x1 = lerp(
        grad(aaa, xf, yf, zf),
        grad(baa, xf - 1, yf, zf), u);
    x2 = lerp(
        grad(aba, xf, yf - 1, zf),
        grad(bba, xf - 1, yf - 1, zf), u);
    double y1 = lerp(x1, x2, v);

    x1 = lerp(
        grad(aab, xf, yf, zf - 1),
        grad(bab, xf - 1, yf, zf - 1), u);
    x2 = lerp(
        grad(abb, xf, yf - 1, zf - 1),
        grad(bbb, xf - 1, yf - 1, zf - 1), u);
    double y2 = lerp(x1, x2, v);

    return (lerp(y1, y2, w) + 1) / 2;
  }

  public int inc(int num) {
    ++num;
    if (repeat > 0) num %= repeat;
    return num;
  }

  public static double grad(int hash, double x, double y, double z) {
    switch (hash & 0xF) {
      case 0x0:
        return x + y;
      case 0x1:
        return -x + y;
      case 0x2:
        return x - y;
      case 0x3:
        return -x - y;
      case 0x4:
        return x + z;
      case 0x5:
        return -x + z;
      case 0x6:
        return x - z;
      case 0x7:
        return -x - z;
      case 0x8:
        return y + z;
      case 0x9:
        return -y + z;
      case 0xA:
        return y - z;
      case 0xB:
        return -y - z;
      case 0xC:
        return y + x;
      case 0xD:
        return -y + z;
      case 0xE:
        return y - x;
      case 0xF:
        return -y - z;
    }
    return 0; // Never happens.
  }

  public static double fade(double t) {
    // Fade function as defined by Ken Perlin. This eases coordinate
    // values so that they will “ease” towards integral values. This
    // ends up smoothing the final output.

    // 6t⁵ − 15t⁴ + 10t³:
    return t * t * t * (t * (t * 6 - 15) + 10);
  }

  public static double lerp(double a, double b, double x) {
    return a + x * (b - a);
  }
}
